### DISCUSSION TITLE
Simulate server-side PLI packet receive (to form keyframe) from JavaScript client

### DISCUSSION THREAD

**Discussion by mholbergerNIMBL** [pion/webrtc#1478] [Q&A]
I am using Pion webserver to stream (from GStreamer UDP pipe) to a javascript media API client in the browser, like in the rtp-to-webRTC example. 

I am generating snapshots from the vp8 packets on the server, but there is delay in finding a keyframe. The current fix for this is to send a PLI packet from the client on the data channel to force the server to form a keyframe. Unfortunately there is no way to force a PLI packet send from the javascript media library APIs. I can generate PLI packets on the serverside and send them on the datachannel, but this has no effect on the server streaming. 

Is there any way to simulate receiving PLI on the serverside? Is there any solution for forcing PLI sending from a javascript browser client?

**Comment by mholbergerNIMBL** [comment]
Is it possible to send a PLI on the datachannel up-stream to GStreamer from the pion server, instead of just client/server?

**Comment by Sean-Der** [comment]
The 'PLI over DataChannel' is something you will need to translate still. In the `OnMessage` from the DataChannel you need to unpack and generate the PLI.

The browser is probably automatically generating PLIs right now. If you read the incoming RTCP packets do you see them?
